Update the CLanguageStandard and CXXLanguageStandard enums

benrimmington opened the pull request:
Generate the enum cases by preprocessing the [LangStandards.def](https://github.com/apple/llvm-project/blob/apple/main/clang/include/clang/Basic/LangStandards.def) file.

### Motivation:

Follow-up to: apple/swift-package-manager#2716

### Modifications:

* Add enum cases for C17/C18, C2x, and C++20.
* Deprecate enum cases for C++1z.
* Remove enum cases for C++2a (which aren't in [release/5.3][] yet).
* Update the raw string value for C94 (bug fix).

### Result:

* New documentation comments.
* New cases for the latest C/C++ language standards.

[release/5.3]: <https://github.com/apple/swift-package-manager/blob/release/5.3/Sources/PackageDescription/LanguageStandardSettings.swift>

benrimmington commented:
The enum cases were generated with:

https://gist.github.com/benrimmington/3ca4f22389186766c060d99f05784c1a

1.  In the first commit (using `shouldGroupByPrefix = false`), the two C++1z cases were manually moved, to preserve the existing order.

2.  In the second commit (using `shouldGroupByPrefix = true`), [library evolution](https://swift.org/blog/library-evolution/) and the synthesized `Encodable` conformance should allow enum cases to be re-ordered, if readability is improved.

The new C17/C18 modes don't seem to differ from C11, but the guiding principle of the [SE-0181 review](https://forums.swift.org/t/review-se-0181-package-manager-c-c-language-standard-support/6306/6) was "supporting all of the non-deprecated GCC/Clang modes".

For the deprecated C++1z modes, is the `introduced: 4` correct?

Cc: @ddunbar

abertelrud commented:
@swift-ci please smoke test

abertelrud commented on Sources/PackageDescription/LanguageStandardSettings.swift:
```diff
@@ -10,66 +10,126 @@
 public enum CLanguageStandard: String, Encodable {
-    /// The identifier for the C89 language standard.
+
+    /// ISO C 1990.
```
Should this be ISO C 1989?

abertelrud commented on Sources/PackageDescription/LanguageStandardSettings.swift:
```diff
@@ -10,66 +10,126 @@
+    @available(_PackageDescription, introduced: 999.0)
+    case c17
+
+    /// ISO C 2017.
```
Is `c18` a synonym for `c17` or should this be "ISO C 2018"?

abertelrud commented on Sources/PackageDescription/LanguageStandardSettings.swift:
```diff
@@ -10,66 +10,126 @@
+    case c2x
+
+    /// ISO C 1990 with GNU extensions.
     case gnu89
```
Should this be "ISO C 1989"?  In general, do the years in the enum cases match the years of the ISO standards, or are there differences?

abertelrud commented on Sources/PackageDescription/LanguageStandardSettings.swift:
```diff
@@ -10,66 +10,126 @@
+    @available(_PackageDescription, introduced: 999.0)
+    case gnu17
+
+    /// ISO C 2017 with GNU extensions.
```
Same question about 2017 vs 2018.

abertelrud reviewed (commented):
First, thanks for this clean-up!

Several of the enum cases have the same comments.  From the `.def` file it looks as if they are synonyms... should it perhaps be noted in the comments that this is intentional (that it's a synonym?).

abertelrud commented:
Looks as if the self-hosted Linux test fails because of package-collection related unit tests:
```
20:42:01 Test Case 'GitHubPackageMetadataProviderTests.testAPILimit' started at 2020-12-02 04:41:33.814
20:42:01 
/home/buildnode/jenkins/workspace/swift-package-manager-self-hosted-Linux-smoke-test/branch-main/swiftpm/Tests/PackageCollectionsTests/GitHubPackageMetadataProviderTests.swift:217: error: GitHubPackageMetadataProviderTests.testAPILimit : XCTAssertEqual failed: ("Optional(PackageCollections.GitHubPackageMetadataProvider.Errors.invalidResponse(https://api.github.com/repos/octocat/Hello-World, "Empty body"))") is not equal to ("Optional(PackageCollections.GitHubPackageMetadataProvider.Errors.apiLimitsExceeded(https://api.github.com/repos/octocat/Hello-World, 5))") - 
20:42:01 
/home/buildnode/jenkins/workspace/swift-package-manager-self-hosted-Linux-smoke-test/branch-main/swiftpm/Tests/PackageCollectionsTests/GitHubPackageMetadataProviderTests.swift:217: error: GitHubPackageMetadataProviderTests.testAPILimit : XCTAssertEqual failed: ("Optional(PackageCollections.GitHubPackageMetadataProvider.Errors.invalidResponse(https://api.github.com/repos/octocat/Hello-World, "Empty body"))") is not equal to ("Optional(PackageCollections.GitHubPackageMetadataProvider.Errors.apiLimitsExceeded(https://api.github.com/repos/octocat/Hello-World, 5))") - 
20:42:01 
/home/buildnode/jenkins/workspace/swift-package-manager-self-hosted-Linux-smoke-test/branch-main/swiftpm/Tests/PackageCollectionsTests/GitHubPackageMetadataProviderTests.swift:217: error: GitHubPackageMetadataProviderTests.testAPILimit : XCTAssertEqual failed: ("Optional(PackageCollections.GitHubPackageMetadataProvider.Errors.invalidResponse(https://api.github.com/repos/octocat/Hello-World, "Empty body"))") is not equal to ("Optional(PackageCollections.GitHubPackageMetadataProvider.Errors.apiLimitsExceeded(https://api.github.com/repos/octocat/Hello-World, 5))") - 
```
@tomerd or @yim-lee  Do these look familiar to you?

tomerd commented:
> @tomerd or @yim-lee Do these look familiar to you?

hmm, I did run into this once before but thought I fixed it. this may be a flaky test. re-run the tests and see if this goes away. if so, I will need to come back to fix this test

benrimmington commented on Sources/PackageDescription/LanguageStandardSettings.swift:
```diff
@@ -10,66 +10,126 @@
 
 /// The supported C language standard to use for compiling C sources in the package.
 public enum CLanguageStandard: String, Encodable {
```
@abertelrud Should I expand the `CLanguageStandard` documentation comment?

```suggestion
/// The supported C language standard to use for compiling C sources in the package.
///
/// Aliases are available for some C language standards. For example,
/// use `c89`, `c90`, or `iso9899_1990` for the "ISO C 1990" standard.
/// To learn more, read the [Clang Compiler User's Manual][1].
///
/// [1]: <https://clang.llvm.org/docs/UsersManual.html#differences-between-various-standard-modes>
public enum CLanguageStandard: String, Encodable {
```

abertelrud replied:
That's great, I think that will be a big help for those who aren't familiar with the naming here and the details of each mode!

benrimmington replied:
@abertelrud After the current smoke tests are finished, I'll commit the suggestions.

benrimmington commented on Sources/PackageDescription/LanguageStandardSettings.swift:
```diff
@@ -10,66 +10,126 @@
+    case iso9899_2018 = "iso9899:2018"
 }
 
 /// The supported C++ language standards to use for compiling C++ sources in the package.
 public enum CXXLanguageStandard: String, Encodable {
```
```suggestion
/// The supported C++ language standard to use for compiling C++ sources in the package.
///
/// Aliases are available for some C++ language standards. For example,
/// use `cxx98` or `cxx03` for the "ISO C++ 1998 with amendments" standard.
/// To learn more, read the [C++ Support in Clang][1] status page.
///
/// [1]: <https://clang.llvm.org/cxx_status.html>
public enum CXXLanguageStandard: String, Encodable {
```

abertelrud commented:
@swift-ci please smoke test

abertelrud commented:
Thanks @benrimmington!  It does look as if one of the unit test failures is related:
```
02:06:19 Test Case '-[PackageLoadingTests.PackageDescription4LoadingTests testLanguageStandards]' started.
02:06:19 
/Users/buildnode/jenkins/workspace/swift-package-manager-PR-osx-smoke-test/branch-main/swiftpm/Tests/PackageLoadingTests/PD4LoadingTests.swift:325: error: -[PackageLoadingTests.PackageDescription4LoadingTests testLanguageStandards] : XCTAssertEqual failed: ("Optional("iso9899:199409")") is not equal to ("Optional("iso9899:1994")")
```

benrimmington commented:
@abertelrud I've updated the test assertion and documentation comments, and squashed the commits.

I didn't see the test failure locally, because it only gets as far as the guard statement.

https://github.com/apple/swift-package-manager/blob/c68fd5b22b51da157a93eaaa4aaa0e4d90bc38d9/Tests/PackageLoadingTests/PD4LoadingTests.swift#L308-L309

abertelrud commented:
@swift-ci please smoke test

abertelrud commented:
Great, thanks!

abertelrud commented:
@swift-ci please smoke test
